### ISSUE TITLE
Using [key: string] in an interface doesn't force string type when using keyof on the interface

### ISSUE THREAD

**Issue by dwilt** [microsoft/TypeScript#44598]
# Bug Report

### 🔎 Search Terms

`keyof interface index`

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about `index`

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ssl=2&ssc=11&pln=2&pc=17#code/KYDwDg9gTgLgBASwHY2FAZgQwMbDgBTQFsEBnUhCJUgZRigFdsYGo8BvAKDjgG1kAJqABccUvWQBzALqiARhAgAbYJiQBuTgF9OnbFXFiIRYDAAWU0QGtgATwjoCxMhQN1GzVngC8cAIyaegbKwAB0ShCSABSkxqYWSJIAlOpAA)

### 💻 Code

```ts
export interface PermissionsStructure {
  [index: string]: boolean;
}

const something: keyof PermissionsStructure = 1;
```

### 🙁 Actual behavior

TS doesn't complain that we're trying to assign a number to a string

### 🙂 Expected behavior

I would expect TS to enforce that when you use `keyof` of an interface who has a defined index type, that it gets checked

**Comment by RyanCavanaugh** [comment]
When an object has a string index signature, its `keyof` is `string | number`. This is because it's legal to write
```ts
const m = somePermissionsStructure[0];
```
which is legal because this is exactly the same as writing
```ts
const m = somePermissionsStructure["0"];
```
